src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -535,14 +540,24 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
 
                 InitializeStreams(messageBody);
 
-                var httpContext = application.CreateContext(this);
+                // Initialize the HttpContext before we call into the IHttpApplication
+                if (_httpContext is null)
```

Comment:
If this is a step too far I can revert this commit and go back to this one https://github.com/aspnet/AspNetCore/pull/6424/commits/8526622fefd26f95528ecf9127ce6a9b7c7f1315. The difference is that it would lazily initialize the http context in the rare case where the IHttpContextFactory is overridden by the user. It means this field would always be null instead of being assigned and those extra calls would be avoided.